src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -229,152 +299,653 @@ ValidatorList::CacheValidatorFile(
+                    info.manifest = blobInfo[jss::manifest].asString();
+                }
+            }
+            assert(result.size() == blobs.size());
+            return result;
+        }
+    }
+}
+
+// static
+std::vector<ValidatorBlobInfo>
+ValidatorList::parseBlobs(protocol::TMValidatorList const& body)
+{
+    return {{body.blob(), body.signature(), {}}};
+}
+
+// static
+std::vector<ValidatorBlobInfo>
+ValidatorList::parseBlobs(protocol::TMValidatorListCollection const& body)
+{
+    if (body.blobs_size() > MAX_SUPPORTED_BLOBS)
+        return {};
+    std::vector<ValidatorBlobInfo> result;
+    result.reserve(body.blobs_size());
+    for (auto const& blob : body.blobs())
+    {
+        ValidatorBlobInfo& info = result.emplace_back();
+        info.blob = blob.blob();
+        info.signature = blob.signature();
+        if (blob.has_manifest())
+        {
+            info.manifest = blob.manifest();
+        }
+    }
+    assert(result.size() == body.blobs_size());
+    return result;
+}
+
+std::size_t
+splitMessageParts(
+    std::vector<ValidatorList::MessageWithHash>& messages,
+    protocol::TMValidatorListCollection const& largeMsg,
+    std::size_t maxSize,
+    std::size_t begin,
+    std::size_t end);
+
+std::size_t
+splitMessage(
+    std::vector<ValidatorList::MessageWithHash>& messages,
+    protocol::TMValidatorListCollection const& largeMsg,
+    std::size_t maxSize,
+    std::size_t begin = 0,
+    std::size_t end = 0)
+{
+    if (begin == 0 && end == 0)
+        end = largeMsg.blobs_size();
+    assert(begin < end);
+    if (end <= begin)
+        return 0;
+
+    auto mid = (begin + end) / 2;
+    // The parts function will do range checking
+    // Use two separate calls to ensure deterministic order
+    auto result = splitMessageParts(messages, largeMsg, maxSize, begin, mid);
+    return result + splitMessageParts(messages, largeMsg, maxSize, mid, end);
+}
+
+std::size_t
+splitMessageParts(
+    std::vector<ValidatorList::MessageWithHash>& messages,
+    protocol::TMValidatorListCollection const& largeMsg,
+    std::size_t maxSize,
+    std::size_t begin,
+    std::size_t end)
+{
+    if (end <= begin)
+        return 0;
+    if (end - begin == 1)
+    {
+        protocol::TMValidatorList smallMsg;
+        smallMsg.set_version(1);
+        smallMsg.set_manifest(largeMsg.manifest());
+
+        auto const& blob = largeMsg.blobs(begin);
+        smallMsg.set_blob(blob.blob());
+        smallMsg.set_signature(blob.signature());
+        // This is only possible if "downgrading" a v2 UNL to v1.
+        if (blob.has_manifest())
+            smallMsg.set_manifest(blob.manifest());
+
+        auto message =
+            std::make_shared<Message>(smallMsg, protocol::mtVALIDATORLIST);
+
+        if (message->getBuffer(compression::Compressed::Off).size() <= maxSize)
+        {
+            messages.emplace_back(message, sha512Half(smallMsg), 1);
+            return messages.back().numVLs;
+        }
```

Comment:
Suppose we land in the "implied" else clause (the message was too large) for some of the _VLs_ in a message, while other _VLs_ from the same original message were successfully appended to `messages`. In this case we can end up sending an incomplete collection of validator lists. Is this a scenario we care about? Based on the implementation it seems that we're opting to silently send an incomplete message, rather than failing loudly, or somehow denoting the deficiency in the outgoing message.

Reply:
This is true. I'm going to sidestep your question, and I'm just going to send any v1 messages regardless of their size.

The current UNL is about 20Kb. The message size limit is 64Mb. I think I was being overly paranoid when I spec'd this out in the first place, because to come anywhere near that, the UNL file size would have to increase over 3000x. I'm tempted to remove the `splitMessage` logic entirely, but I'll leave it for now, and have these v1 messages send regardless.